common/src/main/java/org/apache/nemo/common/DirectByteBufferOutputStream.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.nemo.common;
+
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.util.LinkedList;
+
+/**
+ * This class is a customized output stream implementation backed by
+ * {@link ByteBuffer}, which utilizes off heap memory when writing the data.
+ * Memory is allocated when needed by the specified {@code pageSize}.
+ */
+public class DirectByteBufferOutputStream extends OutputStream {
+
+  private final LinkedList<ByteBuffer> dataList = new LinkedList<>();
+  private final int pageSize;
+
+  /**
+   * Default constructor.
+   * Sets the {@code pageSize} as default size of 4KB.
+   */
+  public DirectByteBufferOutputStream() {
+    pageSize = 4096;
+  }
+
+  /**
+   * Constructor specifying the {@code pageSize}.
+   *
+   * @param size should be a power of 2 and greater than 4KB.
+   */
+  public DirectByteBufferOutputStream(final int size) {
+    if (size < 4096 || (size & (size - 1)) != 0) {
+      throw new IllegalArgumentException("Invalid pageSize");
+    }
+    pageSize = size;
+  }
+
+  /**
+   * Allocates new {@link ByteBuffer} with the capacity equal to
+   * {@code pageSize}.
+   */
+  private void newLastBuffer() {
+    dataList.addLast(ByteBuffer.allocateDirect(pageSize));
+  }
+
+  /**
+   * Writes the specified byte to this output stream.
+   *
+   * @param   b   the byte to be written.
+   */
+  @Override
+  public void write(final int b) {
+    ByteBuffer currentBuf = (dataList.isEmpty() ? null : dataList.getLast());
```

Comment:
The null only happens at the first time, is this correct? then, why don't you just create a byteBuf in constructor?

Reply:
Also, it would be good to define the currentBuf as an instance variable

Reply:
This is related to the `close()` method that it can be called even after the output stream is closed. Depending on the `close()` issue, I will make changes to this. It will be much better. Thank you.

Reply:
For this issue, before actually writing data using `DirectByteBufferOutputStream` we might not want to take up the memory, so I guess it'd be okay to check null at first write :)